overlord/patch/patch4.go

```diff
@@ -0,0 +1,261 @@
+}
+
+func (ss *patch4SnapSetup) Name() string {
+	if ss.SideInfo.RealName == "" {
+		panic("SnapSetup.SideInfo.RealName not set")
+	}
+	return ss.SideInfo.RealName
+}
+
+func (ss *patch4SnapSetup) Revision() snap.Revision {
+	return ss.SideInfo.Revision
+}
+
+type patch4SnapState struct {
+	SnapType string            `json:"type"` // Use Type and SetType
+	Sequence []*patch4SideInfo `json:"sequence"`
+	Active   bool              `json:"active,omitempty"`
+	Current  snap.Revision     `json:"current"`
+	Channel  string            `json:"channel,omitempty"`
+	Flags    patch4Flags       `json:"flags,omitempty"`
+}
+
+func (snapst *patch4SnapState) LastIndex(revision snap.Revision) int {
+	for i := len(snapst.Sequence) - 1; i >= 0; i-- {
+		if snapst.Sequence[i].Revision == revision {
+			return i
+		}
+	}
+	return -1
+}
+
+type patch4T struct{} // for namespacing of the helpers
+
+func (p4 patch4T) taskSnapSetup(task *state.Task) (*patch4SnapSetup, error) {
+	var ss patch4SnapSetup
+
+	switch err := p4.get(task, "snap-setup", true, &ss); err {
+	case state.ErrNoState:
+		// continue below
+	case nil:
+		return &ss, nil
+	default:
+		return nil, err
+	}
+
+	var id string
+	if err := p4.get(task, "snap-setup-task", false, &id); err != nil {
+		return nil, err
+	}
+
+	if err := p4.get(task.State().Task(id), "snap-setup", false, &ss); err != nil {
+		return nil, err
+	}
+
+	return &ss, nil
+}
+
+func (p4 patch4T) snapSetupAndState(task *state.Task) (*patch4SnapSetup, *patch4SnapState, error) {
+	var snapst patch4SnapState
+
+	ss, err := p4.taskSnapSetup(task)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var snaps map[string]*json.RawMessage
+	err = task.State().Get("snaps", &snaps)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot get snaps state: %v", err)
+	}
+	raw, ok := snaps[ss.Name()]
+	if !ok {
+		return nil, nil, fmt.Errorf("cannot get snap state for %q: %v", ss.Name(), err)
+	}
+	err = json.Unmarshal([]byte(*raw), &snapst)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot get state for snap %q: %v", ss.Name(), err)
+	}
+
+	return ss, &snapst, err
+}
+
+func (patch4T) get(task *state.Task, key string, passThroughMissing bool, value interface{}) error {
+	err := task.Get(key, value)
+	if err == nil || (passThroughMissing && err == state.ErrNoState) {
+		return err
+	}
+	change := task.Change()
+
+	return fmt.Errorf("cannot get %q from task %s (%s) of change %s (%s): %v",
+		key, task.ID(), task.Kind(), change.ID(), change.Kind(), err)
+}
+
+func (p4 patch4T) addCleanup(task *state.Task) error {
+	// NOTE we could check for the status of the change itself, but
+	// copy-snap-data is the one creating the trash, so if it's run there's
+	// no sense in fiddling with the change.
+	if task.Status().Ready() {
+		return nil
+	}
+
+	ss, err := p4.taskSnapSetup(task)
+	if err != nil {
+		return err
+	}
+
+	var tid string
+	if err := p4.get(task, "snap-setup-task", false, &tid); err != nil {
```

Comment:
I find the boolean for passthrough a bit difficult to read, I wonder if it would make sense to have two helpers here? one like `get()` that behaves normally and return ErrNoState. And `mustGet` or `getOrBeReallyUnahappy()` or `getErrIfMissing` something? This will also make get() a little easier to read on hot days like today :)